src/Orleans/Utils/IKeyedServiceDescriptor.cs

```diff
@@ -0,0 +1,85 @@
+﻿
+using System;
+using System.Linq;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Orleans
+{
+    public static class KeyedServiceDescriptorExtensions
+    {
+        /// <summary>
+        /// Add supported for keyed services to a service collection
+        /// </summary>
+        public static IServiceCollection SupportKeyedServices(this IServiceCollection collection)
+        {
+            return collection.AddSingleton(typeof(IKeyedServiceCollection<,>), typeof(KeyedServiceCollection<,>));
+        }
+
+        /// <summary>
+        /// Associates a key with a specialization of a service type.
+        /// </summary>
+        public static IServiceCollection WithKey<TKey, TService, TSpecialization>(this IServiceCollection collection, TKey key)
```

Comment:
I just realized you commented on the same line as I did, but weren't responding to my comment. My misunderstanding.  Please consider my comments with the context of my comment on this same line.  The intended pattern was not intended to be:

services.WithKey<string, ISomeService, SomeService>("la"); 

It was intended to be used like:
services.AddTransient< SomeService >.WithKey<string, ISomeService, SomeService>("la");

But as you can see, the api doesn't enforce this usage, just enables it, which I suspected may be unclear, and you're comment has confirmed.